yetanotherco/aligned_layer
# feat(cli): add get-user-first-nonce + get-user-number-queued-proofs cli commands

# Add `get-user-first-nonce` to cli

## Description

Adds a function to the aligned-cli `get-user-first-nonce` to retrieve the users nonce from ethereum. This is useful in scenarios where you need to replace your proofs with higher fees. For that, you'll need the first proof's nonce corresponding to the value in the `BatcherPaymentService` contract.

Adds a function to the aligned-cli  `get-user-number-queued-proofs` to retrieve the number of proofs a user has in the batcher queue.

**To Test:**
- Setup a local devnet.
- Fund a wallet: 
```shell
cast send <YOUR_WALLET_ADDRESS> --private-key <ANVIL_RICH_ACCOUNT> --value 1ether
```
- cd into aligned cli: cd batcher/aligned
- Fund wallet on batcher wallet:
```shell
cargo run --release -- deposit-to-batcher --network devnet --amount 0.1ether --keystore_path <KEYSTORE_PATH>
```
- Send some proofs with a low max fee to the batcher so that they are not submitted:
```shell
cargo run --release -- submit  --proving_system Risc0 --proof ../../scripts/test_files/risc_zero/fibonacci_proof_generator/risc_zero_fibonacci.proof  --vm_program ../../scripts/test_files/risc_zero/fibonacci_proof_generator/fibonacci_id.bin --public_input ../../scripts/test_files/risc_zero/fibonacci_proof_generator/risc_zero_fibonacci.pub --repetitions <BURST_SIZE> --keystore_path <YOUR_KEYSTORE_PATH> --network devnet  --max_fee 10  --nonce "0"
```
- Get the nonce from ethereum:
```shell
cargo run --release -- get-user-nonce-from-ethereum --user_addr <YOUR_WALLET_ADDRESS>
```
You should see it is 0.

- Get the nonce from the Batcher:
```shell
cargo run --release -- get-user-nonce --user_addr <YOUR_WALLET_ADDRESS>
```

You should see it is <BURST_SIZE>

- Get the amount of proofs in batcher queue:
```shell
cargo run --release -- get-amount-of-proofs-in-batcher-queue --user_addr <YOUR_WALLET_ADDRESS>
```

- Bump the fees using the `nonce` you got from the Ethereum:
```shell
cargo run --release -- submit  --proving_system Risc0 --proof ../../scripts/test_files/risc_zero/fibonacci_proof_generator/risc_zero_fibonacci.proof  --vm_program ../../scripts/test_files/risc_zero/fibonacci_proof_generator/fibonacci_id.bin --public_input ../../scripts/test_files/risc_zero/fibonacci_proof_generator/risc_zero_fibonacci.pub --repetitions 3 --keystore_path <YOUR_KEYSTORE_PATH> --network devnet  --max_fee 1000000000000  --nonce "0"
```

## Type of change

Please delete options that are not relevant.


## Checklist

  - [Link to the open issue addressing it]()
